### ISSUE TITLE
How to: pass a user matrix to fit function

### ISSUE THREAD

**Issue by esube** [fchollet/keras#2383]

**Comment by carlthome** [comment]
Could you clarify?

**Comment by esube** [comment]
@carlthome thanks for responding!
If I have a matrix that I want to use in a loss function as input just like sample_weights etc ..., Is there a way to create a place holder for it when the model is compiled and pass it to a fit or similar function. 

In other words, can I pass some matrix (Tensor in general sense) to fit function so I can write a custom cost function that utilizes that input?

**Comment by carlthome** [comment]
This might be inspirational https://github.com/fchollet/keras/issues/2115

**Comment by esube** [comment]
@carlthome Thank a lot for your kind reply.
The partial function idea is good. But, could it work with a variable and not a constant weight matrix?
My weight matrix is calculated from the data on the fly, but, before model compilation.

So, that's why I wanted a way to sneak in placeholder to my matrix at model compilation time as there is no need to re-create a new tensor and fill it with the weight matrix each time the loss function is called.

In short what I want to do is something similar to:

`W = get_weights(y_train) # 2D matrix same as y_train (n_samples x n_labels)`

`def my_loss(y_true, y_pred):`
      `score = K.binary_crossentropy(y_true, y_pred)`
      `score *= W`
      `return K.mean(score, axis=-1)`

But, since W is not a matrix of constants but a variable computed from y_train once, how can I get it to the GPU so I can use its symbolic equivalent in the loss function?

**Comment by carlthome** [comment]
In that case, where the values depend on the current output, we're in the same boat: https://github.com/fchollet/keras/issues/2121

@rpinsler's suggestion of wrapping the loss function in a closure with some numpy slicing should work for me, and might also work for you.

**Comment by esube** [comment]
@carlthome Thank you so much for taking the time to point me to useful related issues.
I ended up creating a graph with K outputs, the outputs being just single binary classifiers and since and I sneaked in my 2D (NxK) weight matrix as K sliced 1D (Nx1) as sample weights to the K outputs of the graph and that seems to do the trick. 

The reason this works is, keras has a weighted_objective(fn) function that basically does what I explained above, but, the main limitation was, with just sequential model, you cannot pass 2D sample weight matrix cos as @fchollet tried to explain in #2408, they want sample_weights to mean just that.

But, another related question: I tried to upgrade to the latest keras and my graphs stopped working. Same Graph model that worked fine with v 0.3.2 stopped working (complaining about unconnected graph or sth) in 1.0.1. So, I had to revert back to 0.3.2. I also noticed that now there's a legacy folder and graph was pointing there. Is graph removed in favor of the new functional API?

**Comment by carlthome** [comment]
Yes, Graph is deprecated.

**Comment by esube** [comment]
Well, that's a bummer! Development in keras is so fast that it is hard to keep up and there's significant re-write every now and then. This seems problematic. I don't want to re-write my graph workaround to get this problem resolved again in new api. It such a lot of work to do. Why is such drastic decisions made within a short period of time? Why not at least keep a common high level interface and change the internals? 

I am sure there's a very good reason and I am sure the new functional API makes it much more easier than Graph. But, breaking user code every now and then in such a significant manner seems a bit harsh.

Btw, you can close this issue.

Thanks once again!

**Comment by carlthome** [comment]
It is deprecated but not broken. See https://github.com/fchollet/keras/issues/2462#issuecomment-213325034
